src/modules/wallet/provider.tsx

```diff
@@ -1,80 +1,92 @@
-import React, { useEffect, useState } from 'react'
-import walletSDK from '@/modules/wallet/client'
-import { useRouter } from 'next/router'
-import { isNil } from 'ramda'
-import { toast } from 'react-toastify'
-import { Wallet } from '@/modules/wallet/types'
-import { Solana } from '@/modules/solana/types'
-import { initArweave } from '@/modules/arweave'
-import arweaveSDK from '@/modules/arweave/client'
-import { WalletContext, WalletContextProps } from './context'
+import { initArweave } from '@/modules/arweave';
+import arweaveSDK from '@/modules/arweave/client';
+import { Solana } from '@/modules/solana/types';
+import walletSDK from '@/modules/wallet/client';
+import { Wallet } from '@/modules/wallet/types';
+import { useRouter } from 'next/router';
+import { isNil } from 'ramda';
+import React, { useState } from 'react';
+import { toast } from 'react-toastify';
+import { WalletContext, WalletContextProps } from './context';
 
 type WalletProviderProps = {
-  wallet?: Wallet,
+  wallet?: Wallet;
   solana?: Solana;
   children: (props: WalletContextProps) => React.ReactElement;
-}
-
+};
 
 const upsertWallet = async (pubkey: string) => {
-  return walletSDK.find(pubkey)
-    .then((wallet: any) => {
-      if (!wallet) {
-        return walletSDK.create(pubkey)
-      }
+  return walletSDK.find(pubkey).then((wallet: any) => {
+    if (!wallet) {
+      return walletSDK.create(pubkey);
+    }
 
-      return Promise.resolve(wallet)
-    })
-}
+    return Promise.resolve(wallet);
+  });
+};
 
 export const WalletProvider = ({ children }: WalletProviderProps) => {
-  const router = useRouter()
-  const arweave = initArweave()
-  const [verifying, setVerifying] = useState(false)
-  const [initializing, setInitialization] = useState(true)
-  const [wallet, setWallet] = useState<Wallet>()
-  const [solana, setSolana] = useState<Solana>()
+  const router = useRouter();
+  const arweave = initArweave();
+  const [verifying, setVerifying] = useState(false);
+  const [initializing, setInitialization] = useState(true);
+  const [wallet, setWallet] = useState<Wallet>();
+  const [solana, setSolana] = useState<Solana>();
 
-  useEffect(() => {
-    if (process.browser) {
-      window.addEventListener("load", () => {
-        setSolana(window.solana)
-        setInitialization(false)
-      }, false)
+  if (typeof window === 'object') {
+    if (window.solana && window.solana?.connect) {
+      if (window.solana !== solana) {
+        setSolana(window.solana);
+        setInitialization(false);
+      }
+    } else {
+      window.addEventListener(
+        'load',
+        () => {
+          setSolana(window.solana);
+          setInitialization(false);
+        },
+        { capture: false }
+      );
     }
-  }, [])
+  }
```

Comment:
The bulk of the changes happen here - the event handler for `load` does not seem to be called very often, so we optimistically look for an updated Phantom every time.